File: py/ndep.py
import numpy
import pandas

import matplotlib
matplotlib.use('agg')
import matplotlib.pyplot as pyplot
import matplotlib.ticker as ticker

import util


Npt, pt = util.set_point_type()
Nls, ls = util.set_line_style()
Ncol, col = util.set_color_palette_for_color_universal_design()


from argparse import ArgumentParser
def get_option():
    argparser = ArgumentParser()
    argparser.add_argument('-c', '--compare',
                           action='store_true',
                           help='Whether to compare with V100 SXM2 (Pascal mode)')
    argparser.add_argument('-p', '--pdf',
                           action='store_true',
                           help='Whether to output figures in PDF format')
    return argparser.parse_args()

# catch optional input arguments
args = get_option()
compare = args.compare
outputPDF = args.pdf
filename = '_ndep'

# embed fonts
pyplot.rcParams['ps.useafm'] = True
pyplot.rcParams['pdf.use14corefonts'] = True
pyplot.rcParams['text.usetex'] = True

# use packages
pyplot.rcParams['text.latex.preamble'] = r'\usepackage{physics,siunitx}'

# specify direction of ticks
pyplot.rcParams['xtick.direction'] = 'in'
pyplot.rcParams['ytick.direction'] = 'in'


# read measured data
particle = 'm31'
num = [1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576, 2097152, 4194304, 8388608, 16777216]
tag = ['001k', '002k', '004k', '008k', '016k', '032k', '064k', '128k', '256k', '512k', '001M', '002M', '004M', '008M', '016M']
series = '.acceleration.block.cc80.'
log = '.time.log'
model = '1410/log/' + particle + '_'
N001k = pandas.read_csv(model + tag[ 0] + series + str(num[ 0]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N002k = pandas.read_csv(model + tag[ 1] + series + str(num[ 1]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N004k = pandas.read_csv(model + tag[ 2] + series + str(num[ 2]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N008k = pandas.read_csv(model + tag[ 3] + series + str(num[ 3]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N016k = pandas.read_csv(model + tag[ 4] + series + str(num[ 4]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N032k = pandas.read_csv(model + tag[ 5] + series + str(num[ 5]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N064k = pandas.read_csv(model + tag[ 6] + series + str(num[ 6]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N128k = pandas.read_csv(model + tag[ 7] + series + str(num[ 7]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N256k = pandas.read_csv(model + tag[ 8] + series + str(num[ 8]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N512k = pandas.read_csv(model + tag[ 9] + series + str(num[ 9]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N001M = pandas.read_csv(model + tag[10] + series + str(num[10]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N002M = pandas.read_csv(model + tag[11] + series + str(num[11]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N004M = pandas.read_csv(model + tag[12] + series + str(num[12]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N008M = pandas.read_csv(model + tag[13] + series + str(num[13]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')
N016M = pandas.read_csv(model + tag[14] + series + str(num[14]) + log, names = ('accuracy_parameter', 'elapsed_total', 'step', 'elapsed_step'), delimiter = '\t')


# pick up representative values
best = [min(N001k['elapsed_step']), min(N002k['elapsed_step']), min(N004k['elapsed_step']), min(N008k['elapsed_step']), min(N016k['elapsed_step']), min(N032k['elapsed_step']), min(N064k['elapsed_step']), min(N128k['elapsed_step']), min(N256k['elapsed_step']), min(N512k['elapsed_step']), min(N001M['elapsed_step']), min(N002M['elapsed_step']), min(N004M['elapsed_step']), min(N008M['elapsed_step']), min(N016M['elapsed_step'])]
worst = [max(N001k['elapsed_step']), max(N002k['elapsed_step']), max(N004k['elapsed_step']), max(N008k['elapsed_step']), max(N016k['elapsed_step']), max(N032k['elapsed_step']), max(N064k['elapsed_step']), max(N128k['elapsed_step']), max(N256k['elapsed_step']), max(N512k['elapsed_step']), max(N001M['elapsed_step']), max(N002M['elapsed_step']), max(N004M['elapsed_step']), max(N008M['elapsed_step']), max(N016M['elapsed_step'])]
median = [numpy.median(N001k['elapsed_step']), numpy.median(N002k['elapsed_step']), numpy.median(N004k['elapsed_step']), numpy.median(N008k['elapsed_step']), numpy.median(N016k['elapsed_step']), numpy.median(N032k['elapsed_step']), numpy.median(N064k['elapsed_step']), numpy.median(N128k['elapsed_step']), numpy.median(N256k['elapsed_step']), numpy.median(N512k['elapsed_step']), numpy.median(N001M['elapsed_step']), numpy.median(N002M['elapsed_step']), numpy.median(N004M['elapsed_step']), numpy.median(N008M['elapsed_step']), numpy.median(N016M['elapsed_step'])]
p1sigma = [numpy.percentile(N001k['elapsed_step'], 84.13), numpy.percentile(N002k['elapsed_step'], 84.13), numpy.percentile(N004k['elapsed_step'], 84.13), numpy.percentile(N008k['elapsed_step'], 84.13), numpy.percentile(N016k['elapsed_step'], 84.13), numpy.percentile(N032k['elapsed_step'], 84.13), numpy.percentile(N064k['elapsed_step'], 84.13), numpy.percentile(N128k['elapsed_step'], 84.13), numpy.percentile(N256k['elapsed_step'], 84.13), numpy.percentile(N512k['elapsed_step'], 84.13), numpy.percentile(N001M['elapsed_step'], 84.13), numpy.percentile(N002M['elapsed_step'], 84.13), numpy.percentile(N004M['elapsed_step'], 84.13), numpy.percentile(N008M['elapsed_step'], 84.13), numpy.percentile(N016M['elapsed_step'], 84.13)]
m1sigma = [numpy.percentile(N001k['elapsed_step'], 15.87), numpy.percentile(N002k['elapsed_step'], 15.87), numpy.percentile(N004k['elapsed_step'], 15.87), numpy.percentile(N008k['elapsed_step'], 15.87), numpy.percentile(N016k['elapsed_step'], 15.87), numpy.percentile(N032k['elapsed_step'], 15.87), numpy.percentile(N064k['elapsed_step'], 15.87), numpy.percentile(N128k['elapsed_step'], 15.87), numpy.percentile(N256k['elapsed_step'], 15.87), numpy.percentile(N512k['elapsed_step'], 15.87), numpy.percentile(N001M['elapsed_step'], 15.87), numpy.percentile(N002M['elapsed_step'], 15.87), numpy.percentile(N004M['elapsed_step'], 15.87), numpy.percentile(N008M['elapsed_step'], 15.87), numpy.percentile(N016M['elapsed_step'], 15.87)]
# upper = [numpy.percentile(N001k['elapsed_step'], 75.0), numpy.percentile(N002k['elapsed_step'], 75.0), numpy.percentile(N004k['elapsed_step'], 75.0), numpy.percentile(N008k['elapsed_step'], 75.0), numpy.percentile(N016k['elapsed_step'], 75.0), numpy.percentile(N032k['elapsed_step'], 75.0), numpy.percentile(N064k['elapsed_step'], 75.0), numpy.percentile(N128k['elapsed_step'], 75.0), numpy.percentile(N256k['elapsed_step'], 75.0), numpy.percentile(N512k['elapsed_step'], 75.0), numpy.percentile(N001M['elapsed_step'], 75.0), numpy.percentile(N002M['elapsed_step'], 75.0), numpy.percentile(N004M['elapsed_step'], 75.0), numpy.percentile(N008M['elapsed_step'], 75.0), numpy.percentile(N016M['elapsed_step'], 75.0)]
# lower = [numpy.percentile(N001k['elapsed_step'], 25.0), numpy.percentile(N002k['elapsed_step'], 25.0), numpy.percentile(N004k['elapsed_step'], 25.0), numpy.percentile(N008k['elapsed_step'], 25.0), numpy.percentile(N016k['elapsed_step'], 25.0), numpy.percentile(N032k['elapsed_step'], 25.0), numpy.percentile(N064k['elapsed_step'], 25.0), numpy.percentile(N128k['elapsed_step'], 25.0), numpy.percentile(N256k['elapsed_step'], 25.0), numpy.percentile(N512k['elapsed_step'], 25.0), numpy.percentile(N001M['elapsed_step'], 25.0), numpy.percentile(N002M['elapsed_step'], 25.0), numpy.percentile(N004M['elapsed_step'], 25.0), numpy.percentile(N008M['elapsed_step'], 25.0), numpy.percentile(N016M['elapsed_step'], 25.0)]


fig_box, ax_box, fs_box, ms_box, lw_box = util.set_figure()
if compare:
    v100 = pandas.read_csv('v100_sxm2.csv')
    ax_box[0].plot(v100['Ntot'], v100['elapsed_step'], linestyle = ls[1], linewidth = lw_box, color = col[0], label = 'V100 (SXM2)')

    fig_ratio, ax_ratio, fs_ratio, ms_ratio, lw_ratio = util.set_figure()
    ax_ratio[0].axhline(1555.0 / 900.0, linestyle = ls[2], linewidth = lw_ratio, color = col[0], label = 'memory bandwidth')
    ax_ratio[0].axhline(((64 * 108) * 1.410 * 2) / ((64 * 80) * 1.530 * 2), linestyle = ls[1], linewidth = lw_ratio, color = col[0], label = 'peak performance')
    ax_ratio[0].plot(num, v100['elapsed_step'][v100['Ntot'] <= 16777216] / best, linestyle = ls[0], linewidth = lw_ratio, color = col[0], label = 'measured')
    ax_ratio[0].set_xlabel(r'$N$', fontsize = fs_ratio)
    ax_ratio[0].set_ylabel('Speed up from V100 (SXM2)', fontsize = fs_ratio)
    ax_ratio[0].set_xlim(util.scale_axis(min(num), max(num), True))
    ax_ratio[0].semilogx()
    ax_ratio[0].grid()
    handles, labels = ax_ratio[0].get_legend_handles_labels()
    ax_ratio[0].legend(handles[::-1], labels[::-1], numpoints = 1, loc = 'best', fontsize = fs_ratio)
    fig_ratio.savefig('fig/' + particle + filename + '_speedup' + '.png', format = 'png', dpi = 100, bbox_inches = 'tight')
    if outputPDF:
        fig_ratio.savefig('fig/' + particle + filename + '_speedup' + '.pdf', format = 'pdf', bbox_inches = 'tight')

ax_box[0].plot(num, median, linestyle = ls[2], linewidth = lw_box, color = col[2], label = 'median')
ax_box[0].plot(num, best, linestyle = ls[0], linewidth = lw_box, color = col[1], label = 'fastest')
style_box = dict(lw = 0.75 * lw_box)
style_whi = dict(markersize = ms_box)
style_med = dict(color = col[2], lw = 0.5 * lw_box)
style_fli = dict(marker = 'x', markersize = 0.5 * ms_box)
ax_box[0].boxplot([N001k['elapsed_step'], N002k['elapsed_step'], N004k['elapsed_step'], N008k['elapsed_step'], N016k['elapsed_step'], N032k['elapsed_step'], N064k['elapsed_step'], N128k['elapsed_step'], N256k['elapsed_step'], N512k['elapsed_step'], N001M['elapsed_step'], N002M['elapsed_step'], N004M['elapsed_step'], N008M['elapsed_step'], N016M['elapsed_step']], positions = num, widths = numpy.array(num) * 0.25, boxprops = style_box, whiskerprops = style_whi, medianprops = style_med, flierprops = style_fli)

ax_box[0].set_xlabel(r'$N$', fontsize = fs_box)
ax_box[0].set_ylabel(r'$t_\mathrm{step}$~(\si{s})', fontsize = fs_box)
ax_box[0].set_xlim(util.scale_axis(min(num), max(num), True))

ax_box[0].loglog()
ax_box[0].grid()

handles, labels = ax_box[0].get_legend_handles_labels()
ax_box[0].legend(handles[::-1], labels[::-1], numpoints = 1, loc = 'best', fontsize = fs_box)



# save figures
fig_box.savefig('fig/' + particle + filename + '_box' + '.png', format = 'png', dpi = 100, bbox_inches = 'tight')
if outputPDF:
    fig_box.savefig('fig/' + particle + filename + '_box' + '.pdf', format = 'pdf', bbox_inches = 'tight')

pyplot.close('all')

File: jl/plot/ndep.jl
using ArgParse
function parse_cmd()
    cfg = ArgParseSettings()
    @add_arg_table cfg begin
        # "--target", "-t"
        # help = "filename of the target files"
        # arg_type = String
        # default = ""
        "--speedup", "-s"
        help = "show speed-up ratio"
        action = :store_true
        "--pdf", "-p"
        help = "generate figure in PDF format"
        action = :store_true
    end
    return parse_args(cfg)
end

using FilePaths
using Glob
function get_file_list(folder::String, series::String, tag::String)
    # list = glob(string(folder, series, "*acceleration.block.", tag, ".8388608.time.log"))
    list = glob(string(folder, series, "*acceleration.block.", tag, ".*.time.log"))
    # println(list)
    # println(glob(string(folder, "*acceleration*")))
    # println(length(list))
    return list
end

using DataFrames
using CSV
function read_tsv(file::String)
    tsv = CSV.read(file, DataFrame; delim = "\t", header = ["dacc", "sec", "step", "sec_per_step"], comment = "#")
    return tsv
end

function read_csv(file::String)
    csv = CSV.read(file, DataFrame; delim = ",", comment = "#")
    # println(csv)
    # println(csv.Ntot)
    # println(csv.elapsed_step)
    return csv.Ntot, csv.elapsed_step
end


function get_measured_results(series::String, path::String, gen::String)
    # read measured results
    list = get_file_list(path, series, gen)
    result = read_tsv(list[begin])
    for ii in 2:length(list)
        result = vcat(result, read_tsv(list[ii]))
    end
    # # println(result)
    # # println(result[!, "dacc"])
    # # println(result[result.dacc .== 0.5, :])
    # # println(result[result.dacc .== 0.5, :].sec_per_step)
    dacc = sort(unique(result.dacc), rev = true)
    # # println(dacc)

    time = Array{Float64, 1}(undef, length(dacc))
    for ii in 1:length(dacc)
        data = result[result.dacc .== dacc[ii], :].sec_per_step
        time[ii] = minimum(data)
    end
    # println(time)
    # return dacc, time
    return time
end


using PyPlot
include("../util/pyplot.jl")


function main()
    # read options
    argv = parse_cmd()
    output_pdf = argv["pdf"]
    show_ratio = argv["speedup"]

    # find the latest series of simulation results
    # series = argv["target"] == "" ? util_find.latest_series() : argv["target"]
    series = "m31"

    # initialize matplotlib
    util_pyplot.config()
    fig = util_pyplot.set_Panel()

    # p100_dacc, p100_time = get_measured_results(series, "../../../a100/201225gothic/adep/past/", "cc60")
    # # v100v_dacc, v100v_time = get_measured_results(series, "../../../a100/201225gothic/adep/past/", "cc70")
    # v100p_dacc, v100p_time = get_measured_results(series, "../../../a100/201225gothic/adep/past/", "cc70_60")
    # a100p_dacc, a100p_time = get_measured_results(series, "../../../a100/201225gothic/adep/1410/log/", "cc80")
    # a100s_dacc, a100s_time = get_measured_results(series, "log/", "cc80")
    # # println(a100_dacc)
    # # println(a100_time)

    num = [1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576, 2097152, 4194304, 8388608, 16777216]
    a100s = get_measured_results(string(series, "_N001k"), "log/", "cc80")
    a100s = vcat(a100s, get_measured_results(string(series, "_N002k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N004k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N008k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N016k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N032k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N064k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N128k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N256k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N512k"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N001M"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N002M"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N004M"), "log/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series), "lnk/", "cc80"))
    a100s = vcat(a100s, get_measured_results(string(series, "_N016M"), "log/", "cc80"))

    a100p = get_measured_results(string(series, "_001k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80")
    a100p = vcat(a100p, get_measured_results(string(series, "_002k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_004k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_008k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_016k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_032k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_064k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_128k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_256k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_512k"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_001M"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_002M"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_004M"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_008M"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))
    a100p = vcat(a100p, get_measured_results(string(series, "_016M"), "../../../a100/201225gothic/ndep/1410/log/", "cc80"))

    v100p_Ntot, v100p_time = read_csv("v100_sxm2_ref.csv")


    # fig.ax[begin].plot(p100_dacc, p100_time, util_pyplot.call(fig.point, id = 2), markersize = fig.ms, linestyle = util_pyplot.call(fig.line, id = 3), linewidth = fig.lw, color = util_pyplot.call(fig.color, id = 0), label = "P100 (SXM)")
    # # fig.ax[begin].plot(v100v_dacc, v100v_time, util_pyplot.call(fig.point, id = 1), markersize = fig.ms, linestyle = util_pyplot.call(fig.line, id = 1), linewidth = fig.lw, color = util_pyplot.call(fig.color, id = 2), label = L"V100 (SXM2, compute\_70)", markerfacecolor = "none")
    fig.ax[begin].plot(v100p_Ntot, v100p_time, util_pyplot.call(fig.point, id = 1), markersize = fig.ms, linestyle = util_pyplot.call(fig.line, id = 2), linewidth = fig.lw, color = util_pyplot.call(fig.color, id = 0), label = "V100 (SXM2)")
    fig.ax[begin].plot(num, a100p, util_pyplot.call(fig.point, id = 0), markersize = fig.ms, linestyle = util_pyplot.call(fig.line, id = 1), linewidth = fig.lw, color = util_pyplot.call(fig.color, id = 2), label = "A100 (PCIe)", markerfacecolor = "none")
    fig.ax[begin].plot(num, a100s, util_pyplot.call(fig.point, id = 0), markersize = fig.ms, linestyle = util_pyplot.call(fig.line, id = 0), linewidth = fig.lw, color = util_pyplot.call(fig.color, id = 1), label = "A100 (SXM4)")

    fig.ax[begin].set_xlabel(L"$N_\mathrm{tot}$", fontsize = fig.fs)
    fig.ax[begin].set_ylabel(L"$t_\mathrm{step}~\bqty{\si{s}}$", fontsize = fig.fs)
    fig.ax[begin].grid()
    fig.ax[begin].loglog()

    # add legend
    handles, labels = fig.ax[begin].get_legend_handles_labels()
	fig.ax[begin].legend(reverse(handles), reverse(labels), numpoints = 1, handlelength = 2.0, loc = "best", fontsize = fig.fs)

    # save figures
    fig.fig.savefig(string("fig/", series, "_ndep", ".png"), format = "png", dpi = 100, bbox_inches = "tight")
    if output_pdf
        fig.fig.savefig(string("fig/", series, "_ndep", ".pdf"), format = "pdf", bbox_inches = "tight")
    end

    if show_ratio
        ratio = util_pyplot.set_Panel()
        form = util_pyplot.set_Panel()

        ratio.ax[begin].axhline(1555.0 / 900.0, linestyle = util_pyplot.call(ratio.line, id = 1), linewidth = ratio.lw, color = util_pyplot.call(ratio.color, id = 0), label = "Memory bandwidth")
        ratio.ax[begin].axhline(((64 * 108) * 1.410 * 2) / ((64 * 80) * 1.530 * 2), linestyle = util_pyplot.call(ratio.line, id = 3), linewidth = ratio.lw, color = util_pyplot.call(ratio.color, id = 0), label = "Peak performance")
        ratio.ax[begin].plot(num, v100p_time ./ a100p, util_pyplot.call(ratio.point, id = 1), markersize = ratio.ms, linestyle = util_pyplot.call(ratio.line, id = 2), linewidth = ratio.lw, color = util_pyplot.call(ratio.color, id = 2), label = "A100 (PCIe)")
        ratio.ax[begin].plot(num, v100p_time ./ a100s, util_pyplot.call(ratio.point, id = 0), markersize = ratio.ms, linestyle = util_pyplot.call(ratio.line, id = 0), linewidth = ratio.lw, color = util_pyplot.call(ratio.color, id = 1), label = "A100 (SXM4)")
        form.ax[begin].plot(num, a100p ./ a100s, util_pyplot.call(form.point, id = 0), markersize = form.ms, linestyle = util_pyplot.call(form.line, id = 0), linewidth = form.lw, color = util_pyplot.call(form.color, id = 0))

        ratio.ax[begin].set_xlabel(L"$N_\mathrm{tot}$", fontsize = ratio.fs)
        ratio.ax[begin].set_ylabel("Speed up", fontsize = ratio.fs)
        ratio.ax[begin].grid()
        ratio.ax[begin].semilogx()
        form.ax[begin].set_xlabel(L"$N_\mathrm{tot}$", fontsize = form.fs)
        form.ax[begin].set_ylabel(L"$t_\mathrm{A100 (PCIe)} / t_\mathrm{A100 (SXM4)}$", fontsize = form.fs)
        form.ax[begin].grid()
        form.ax[begin].semilogx()

        # add legend
        handles, labels = ratio.ax[begin].get_legend_handles_labels()
        ratio.ax[begin].legend(reverse(handles), reverse(labels), numpoints = 1, handlelength = 2.0, loc = "best", fontsize = ratio.fs)

        ratio.fig.savefig(string("fig/", series, "_ndep", "_speedup", ".png"), format = "png", dpi = 100, bbox_inches = "tight")
        form.fig.savefig(string("fig/", series, "_ndep", "_form", ".png"), format = "png", dpi = 100, bbox_inches = "tight")
        if output_pdf
            ratio.fig.savefig(string("fig/", series, "_ndep", "_speedup", ".pdf"), format = "pdf", bbox_inches = "tight")
            form.fig.savefig(string("fig/", series, "_ndep", "_form", ".pdf"), format = "pdf", bbox_inches = "tight")
        end

        ratio = nothing
        form = nothing
    end


	fig = nothing
    PyPlot.close("all")
    return nothing

end


main()
